pkg/policyendpoints/manager.go

```diff
@@ -348,3 +239,203 @@ func (m *policyEndpointsManager) getEndpointInfoKey(info policyinfo.EndpointInfo
 	}
 	return hex.EncodeToString(hasher.Sum(nil))
 }
+
+// processExistingPolicyEndpoints processes the existing policies with the coming network policy changes
+// it returned required rules and pod selector changes, and potential modifications and deletions on policy endpoints.
+func (m *policyEndpointsManager) processExistingPolicyEndpoints(
+	policy *networking.NetworkPolicy,
+	existingPolicyEndpoints []policyinfo.PolicyEndpoint, ingressEndpoints []policyinfo.EndpointInfo,
+	egressEndpoints []policyinfo.EndpointInfo, podSelectorEndpoints []policyinfo.PodEndpoint,
+) (
+	map[string]policyinfo.EndpointInfo,
+	map[string]policyinfo.EndpointInfo,
+	sets.Set[policyinfo.PodEndpoint],
+	[]policyinfo.PolicyEndpoint,
+	[]policyinfo.PolicyEndpoint,
+) {
+
+	// Loop through ingressEndpoints, egressEndpoints and podSelectorEndpoints and put in map
+	ingressEndpointsMap := map[string]policyinfo.EndpointInfo{}
+	for _, ingressEndpoint := range ingressEndpoints {
+		ingressEndpointsMap[m.getEndpointInfoKey(ingressEndpoint)] = ingressEndpoint
+	}
+	egressEndpointsMap := map[string]policyinfo.EndpointInfo{}
+	for _, egressEndpoint := range egressEndpoints {
+		egressEndpointsMap[m.getEndpointInfoKey(egressEndpoint)] = egressEndpoint
+	}
+	podSelectorEndpointSet := sets.New[policyinfo.PodEndpoint](podSelectorEndpoints...)
+	// Go over the existing endpoints, and remove entries that are no longer needed
+	var modifiedEndpoints []policyinfo.PolicyEndpoint
+	var potentialDeletes []policyinfo.PolicyEndpoint
+	for i := range existingPolicyEndpoints {
+		ingEndpointList := make([]policyinfo.EndpointInfo, 0, len(existingPolicyEndpoints[i].Spec.Ingress))
+		for _, ingRule := range existingPolicyEndpoints[i].Spec.Ingress {
+			ruleKey := m.getEndpointInfoKey(ingRule)
+			if _, exists := ingressEndpointsMap[ruleKey]; exists {
+				ingEndpointList = append(ingEndpointList, ingRule)
+				delete(ingressEndpointsMap, ruleKey)
+			}
+		}
+		egEndpointList := make([]policyinfo.EndpointInfo, 0, len(existingPolicyEndpoints[i].Spec.Egress))
+		for _, egRule := range existingPolicyEndpoints[i].Spec.Egress {
+			ruleKey := m.getEndpointInfoKey(egRule)
+			if _, exists := egressEndpointsMap[ruleKey]; exists {
+				egEndpointList = append(egEndpointList, egRule)
+				delete(egressEndpointsMap, ruleKey)
+			}
+		}
+		podSelectorEndpointList := make([]policyinfo.PodEndpoint, 0, len(existingPolicyEndpoints[i].Spec.PodSelectorEndpoints))
+		for _, ps := range existingPolicyEndpoints[i].Spec.PodSelectorEndpoints {
+			if podSelectorEndpointSet.Has(ps) {
+				podSelectorEndpointList = append(podSelectorEndpointList, ps)
+				podSelectorEndpointSet.Delete(ps)
+			}
+		}
+		policyEndpointChanged := false
+		if !equality.Semantic.DeepEqual(policy.Spec.PolicyTypes, existingPolicyEndpoints[i].Spec.PodIsolation) {
+			existingPolicyEndpoints[i].Spec.PodIsolation = policy.Spec.PolicyTypes
+			policyEndpointChanged = true
+		}
+
```

Comment:
Let's add a detailed description on what the below code block is trying to accomplish based on what we discussed offline. Will be very helpful for anyone trying to follow this flow..